multi_camera_parallel_grabber/ex-multi_camera_parallel_grabber.cpp

```diff
@@ -0,0 +1,182 @@
+{   // enum for all the trigger types supported by o3d3xx device
+    enum class trigger_mode : int { FREE_RUN = 1, SW = 2, POSITIVE_EDGE = 3, NEGATIVE_EDGE = 4, POSITIVE_N_NEGATIVE = 5 };
+
+    //CHANGE IP addreses to those of camera's avaialible!!!!
+    const auto CAMERA0 = "192.168.0.70"; // Device in software trigger mode
+    const auto CAMERA1 = "192.168.0.71";
+    const auto CAMERA2 = "192.168.0.72";
+
+    //Add the IP of cameras to be used with hardware trigger
+    const std::vector<std::string> camera_ips = {CAMERA1,CAMERA2};
+    int num_cam_finished = camera_ips.size();
+
+    auto j_conf = json::parse(R"(
+         {
+           "ifm3d":
+           {
+             "Device":
+              {
+                "ActiveApplication": "1"
+              },
+              "Apps":
+              [
+                {
+                  "Index": "1",
+                  "TriggerMode":"2",
+                  "LogicGraph": "{\"IOMap\": {\"OUT1\": \"RFT\",\"OUT2\": \"AQUFIN\"},\"blocks\": {\"B00001\": {\"pos\": {\"x\": 200,\"y\": 200},\"properties\": {},\"type\": \"PIN_EVENT_IMAGE_ACQUISITION_FINISHED\"},\"B00002\": {\"pos\": {\"x\": 200,\"y\": 75},\"properties\": {},\"type\": \"PIN_EVENT_READY_FOR_TRIGGER\"},\"B00003\": {\"pos\": {\"x\": 600,\"y\": 75},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT1\"},\"B00005\": {\"pos\": {\"x\": 600,\"y\": 200},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT2\"}},\"connectors\": {\"C00000\": {\"dst\": \"B00003\",\"dstEP\": 0,\"src\": \"B00002\",\"srcEP\": 0},\"C00001\": {\"dst\": \"B00005\",\"dstEP\": 0,\"src\": \"B00001\",\"srcEP\": 0}}}",
+                  "Imager":
+                  {
+                      "ExposureTime": "1000",
+                      "Type":"under5m_moderate",
+                      "FrameRate":"20"
+                  }
+                }
+             ]
+            }
+          }
+      )");
+
+    //for configuration of required trigger on camera
+    void configuration(ifm3d::Camera::Ptr& camera, trigger_mode type)
+    {
+        try
+        {
+            auto application_id = camera->ActiveApplication();
+            j_conf["ifm3d"]["Device"]["ActiveApplication"] = std::to_string(application_id);
+            j_conf["ifm3d"]["Apps"][0]["Index"] = std::to_string(application_id);
+            j_conf["ifm3d"]["Apps"][0]["TriggerMode"] = std::to_string((int)type);
+
+            camera->FromJSON(j_conf);
+        }
+        catch (std::exception &e)
+        {
+            std::cout << e.what();
+        }
+    }
+
+    // for grabbing data when its available
+    void grabimage(ifm3d::FrameGrabber::Ptr& frame_grabber,
+            std::function<void(ifm3d::ImageBuffer::Ptr image_buffer,
+                bool in_time)> callback)
+    {
+        while(true)
+        {
+            std::this_thread::sleep_for(std::chrono::duration<int, std::milli>(1));
+            ifm3d::ImageBuffer::Ptr image_buffer = std::make_shared<ifm3d::ImageBuffer>();
```

Comment:
Why is a new `ImageBuffer` being allocated on each loop iteration? This is inefficient. The internals of the ifm-supported image containers intentionally use the `.create(...)` factory functions on the internally wrapped `cv::Mat` objects so that memory allocations are kept to a minimum in high frequency frame grabbing loops. Writing the code this way will incur expensive heap allocations for every frame acquired from the camera.